### ISSUE TITLE
zfs state advertises to change the system for recordsize with test=True

### ISSUE THREAD

**Issue by dereckson** [saltstack/salt#44404]
### Description of Issue/Question

With test=True, commands should be no op and explain what they should do.

Instead, `zfs.filesystem_present` advertises to update the volume record_size property (unconditionally, even if the volume is already configured like that).

According `zpool history` output, there is no command recorded (and the history logs reconfigure commands too, like a `zfs create -o quux=foo ...` followed by a `zfs set quux=foo ...`, the second will be **logged).**

This is confirmed by the debug log (see below): only `zfs list` and `zfs get` are executed, and no `zfs set`.

So the issue is probably in the reporting.

### Setup

A zfs.filesystem_present state, like:

```
{{ tank }}/mysql-innodb-data:
  zfs.filesystem_present:
    - properties:
        compression: lz4
        recordsize: 16k
        primarycache: metadata
```

### Steps to Reproduce Issue
Apply the state with test=True:

```
$ salt-call --local state.apply path/to/state test=True
          ID: arcology/mysql-innodb-data                                       
    Function: zfs.filesystem_present   
      Result: True                     
     Comment: filesystem arcology/mysql-innodb-data was updated                
     Started: 23:40:57.321366          
    Duration: 143.389 ms               
     Changes:                          
              ----------               
              arcology/mysql-innodb-data:                                      
                  ----------           
                  recordsize:          
                      16k 
$ zpool history
(nothing new)
$ salt-call -l debug --local state.apply path/to/state test=True
...
[DEBUG   ] LazyLoaded zfs.filesystem_present                                                                                                                   
[INFO    ] Running state [arcology/mysql-innodb-data] at time 23:44:51.786683                                                                                  
[INFO    ] Executing state zfs.filesystem_present for [arcology/mysql-innodb-data]                                                                             
[DEBUG   ] zfs.filesystem_present::arcology/mysql-innodb-data::config::create_parent = False
[DEBUG   ] zfs.filesystem_present::arcology/mysql-innodb-data::config::cloned_from = None                                               
[DEBUG   ] zfs.filesystem_present::arcology/mysql-innodb-data::config::properties = OrderedDict([('compression', 'lz4'), ('recordsize', '16k'), ('primarycache'
, 'metadata')])                                                     
[INFO    ] Executing command '/sbin/zfs list -H -t filesystem -o name,used,avail,refer,mountpoint arcology/mysql-innodb-data' in directory '/root'
[DEBUG   ] stdout: arcology/mysql-innodb-data   88K     792G    88K     none
[INFO    ] Executing command '/sbin/zfs get -H -d 1 -o name,property,value compression,recordsize,primarycache arcology/mysql-innodb-data' in directory '/root'
[DEBUG   ] stdout: arcology/mysql-innodb-data   compression     lz4            
arcology/mysql-innodb-data      recordsize      16K                                            
arcology/mysql-innodb-data      primarycache    metadata                                          
[INFO    ] {'arcology/mysql-innodb-data': {'recordsize': '16k'}}
[INFO    ] Completed state [arcology/mysql-innodb-data] at time 23:44:51.940965 duration_in_ms=154.281
...
```

### Versions Report

```
Salt Version:                          
           Salt: 2017.7.1
System Versions:                       
        release: 11.1-RELEASE-p1       
****         system: FreeBSD         
```

**Comment by dereckson** [comment]
By the way, running this without test=True to add a mount point, I've got an extraneous zfs set for the recordsize.

```
2017-11-05.01:45:52 zfs set mountpoint=/var/db/mysql/mysql-innodb-data arcology/mysql-innodb-data
2017-11-05.01:45:52 zfs set recordsize=16k arcology/mysql-innodb-data
```

**Comment by dereckson** [comment]
@sjorge You could be interested by this issue.

**Comment by sjorge** [comment]
I’ll look at this in more detail somewhere next week.

But I think I already see the problem, can you change your state to ‘16K’ with a capital K and try again?

Originally everything was case insensitive for the compared but this got changed a while ago because things like mountpoint, name,... are case sensitive.

Base on the output ZFS is returning 16K while you want 16k... for ZFS it does not matter but for the comparison in purging 16k != 16K.

If this is indeed the case (test mentioned above should confirm this), a fix won’t be be easy... i’ll need to thing about this for a bit. Simply storing a list of case insensitive variables could work, but is ugly and hard to maintain.... 🤔

**Comment by sjorge** [comment]
*purging -> python... on mobile so can’t edit the comment

**Comment by sjorge** [comment]
Adding the -p flag for parsable output might work!
But that would still require a list of properties to manually handle and convert from the human readable to byte notation... still not ideal but probably better in the long run...

```
- SmartOS (build: 20171026T003127Z)
[root@carbon ~]# zfs get -H -o name,property,value recordsize data
data    recordsize      128K
[root@carbon ~]# zfs get -H -p -o name,property,value recordsize data
data    recordsize      131072
[root@carbon ~]# ```

**Comment by dereckson** [comment]
Indeed for the K:

```
         ID: arcology/mysql-innodb-data                                                                                                                       
    Function: zfs.filesystem_present                                                                                                                           
      Result: True                                                                                                                                             
     Comment: filesystem arcology/mysql-innodb-data is up to date                                                                                              
     Started: 18:08:16.078113                                                                                                                                  
    Duration: 138.717 ms                                                                                                                                       
     Changes: 
```

**Comment by sjorge** [comment]
Thanks for confirming, I know what the issues is now... still thinking about how to best fix it.

**Comment by sjorge** [comment]
Some ideas from ZFS people on twitter:

```@redtwitdown‬ 06/11/2017, 06:25: @sjorge‬ zfs get |& fgrep '<size>' *slightly* better than parsing manpage?```

**Comment by sjorge** [comment]
This will have to wait till after #44371 and #44380 get merged, trying to avoid needless and confusing rebases.

**Comment by lordcirth** [comment]
Similar issue, not sure if the same bug or not:

`zfs_filesystem:          
  zfs.filesystem_present:
    - name: 'tank/data'  
    - properties:        
        copies: 2        
`
When run with test=True, it displays blue (Changed) but doesn't actually change:
`----------                                    
          ID: zfs_filesystem                  
    Function: zfs.filesystem_present          
        Name: tank/data                       
      Result: True                            
     Comment: filesystem tank/data was created
     Started: 13:24:24.462470                 
    Duration: 10.077 ms                       
     Changes:                                 
              ----------                      
              tank/data:                      
                  ----------                  
                  copies:                     
                      2                       
`

When run without test=True, it sets copies=2, but continues to claim it changed something when run again.

**Comment by sjorge** [comment]
Can you post a debug log like in the first comment? It could help me determine that.

**Comment by lordcirth** [comment]
With test=True:
[zfs_debug_test.txt](https://github.com/saltstack/salt/files/1451795/zfs_debug_test.txt)

Without:
[zfs_debug.txt](https://github.com/saltstack/salt/files/1451794/zfs_debug.txt)

**Comment by sjorge** [comment]
This might be a different cause but more or less the same issues with the value compare.

Try quoting the 2 in your state file. I think it might trip up because it is comparing 2 (int) to 2 (str).

**Comment by lordcirth** [comment]
@sjorge good idea, '2' as a string does report 'Clean'.  Perhaps the code should cast to string?

**Comment by sjorge** [comment]
Once #44371 is merged I am going to attempt to parse the types from the 'zfs get' (hopefully)
Then I can cast to str or int depending on the type. That should hopefully fix this.

The same should probably probably also exists for the zpool properties so I must do the same for those.

**Comment by sjorge** [comment]
OK, there seems to be an easier way... still waiting on the other two PR's first but here is some notes/POC.

```python
[.]$ python
Python 2.7.10 (default, Jul 15 2017, 17:16:57)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> num = "123"
>>> siz = "128k"
>>> str = "hello"
>>> print str,siz,num
hello 128k 123
>>> num == 123
False
>>> num.isdigit()
True
>>> siz.isdigit()
False
>>> i_num = float(num) if num.isdigit() else num
>>> i_num == 123
True
>>> siz.upper()
'128K'
>>> import re
>>> x = re.match("^([0-9]*)([A-Z]*)$", siz.upper())
>>> x.group(0)
'128K'
>>> x.group(1)
'128'
>>> x.group(2)
'K'
>>> scale = {'K': 1024, 'M': 1024 * 1024, 'G': 1024 * 1024 * 1024, 'T': 1024 * 1024 * 1024 * 1024 }
>>> int(x.group(1)) * scale[x.group(2)]
131072
```

A simple translate function that is called on the output of zfs/zpool should do the trick.

1. str.isdigit() will detect numbers, we can always convert those to a float.
```python
## NOTE: this is true
int("123") == 123

## NOTE: this is also true true
float("123.0") == 123

## NOTE: same for this
float("123.0") == float(123)
```

This does still leave detecting of a float value... perhaps using a regex instead of str.isdigit() can be used.

2.  a simple precompiled regex should be able to detect the size.

This regex needs to take into account that the number can be an *int OR a float*
The suffix (case insensitive) can be K(B), M(B), G(B), T(B), P(B), E(B), Z(B) (Yottabytes are not supported)

**Comment by sjorge** [comment]
Small update, I haven't forgotten about this, still waiting for those two other PR's get go in first to avoid confusing rebasing and cherry picking.

**Comment by sjorge** [comment]
This should have been fix in early December, by PR #44796
